### PR TITLE
C++: Fix AV Rule 85

### PR DESCRIPTION
**Pull request by ian-semmle**
We have to be careful to avoid giving alerts to functions that might be correctly defined, but we can't see the definition as they weren't instantiated.

### CODE REVIEW

File: cpp/ql/test/query-tests/jsf/4.10 Classes/AV Rule 85/AV Rule 85.expected [hunk 1]
```diff
@@ -3,39 +3,7 @@
...
-| AV Rule 85.cpp:79:7:79:14 | MyClass8<T> | When two operators are opposites, both should be defined and one should be defined in terms of the other. Operator operator<= is declared on line 90, but it is not defined in terms of its opposite operator operator>. |
-| AV Rule 85.cpp:79:7:79:14 | MyClass8<T> | When two operators are opposites, both should be defined and one should be defined in terms of the other. Operator operator> is declared on line 88, but it is not defined in terms of its opposite operator operator<=. |
-| AV Rule 85.cpp:79:7:79:14 | MyClass8<T> | When two operators are opposites, both should be defined and one should be defined in terms of the other. Operator operator>= is declared on line 85, but it is not defined in terms of its opposite operator operator<. |
-| AV Rule 85.cpp:79:7:79:14 | MyClass8<int> | When two operators are opposites, both should be defined and one should be defined in terms of the other. Operator operator< is declared on line 83, but it is not defined in terms of its opposite operator operator>=. |
```
Comment (jonas-semmle) [inline cpp/ql/test/query-tests/jsf/4.10 Classes/AV Rule 85/AV Rule 85.expected]: This doesn't look right. Why do the alerts for `MyClass{5,6,7}` go away completely? The comments in the test source say they should still be there. And why is there no alert added for the new `MyClass10`?

**Comment by ian-semmle** [comment]
We don't get bodies for the templates, so we can't give alerts based on them.

**Comment by jonas-semmle** [comment]
I'm still confused by this PR and by this query. The test seems to have produced alerts for code marked as `// GOOD` since it was introduced in 174f1fe1c197e88220efa6b99fcece06cfacd199.

Is it correct that if we get the body for an instantiation then we also get the body of its template? If that's the case, we should be able to ignore instantiations in this query. That seems a lot easier than trying to match instantiated operators with their templates by matching on the name. Matching by name is fragile in any case as operators can be overloaded. I'd much appreciate some QLDoc about which declarations and bodies are guaranteed to get extracted, especially now that it's changed after `discover_walk`.

Ignoring instantiations doesn't by itself fix the problems with this query because we also need to suppress alerts from operators whose body did not get extracted. This might mean we'll end up suppressing all the same true positives that you're suppressing in this PR, but it should still make the query less complicated.

File: cpp/ql/test/query-tests/jsf/4.10 Classes/AV Rule 85/AV Rule 85.expected (hunk 1 above)
Comment (geoffw0) [reply cpp/ql/test/query-tests/jsf/4.10 Classes/AV Rule 85/AV Rule 85.expected]: Perhaps we should modify the test to instantiate these functions, and add just one test case where the function is never instantiated to document that behaviour.  This way the tests will be providing coverage for whatever issues they were originally intended to.

File: cpp/ql/src/jsf/4.10 Classes/AV Rule 85.ql [hunk 2]
```diff
@@ -28,14 +28,34 @@ predicate implementedAsNegationOf(Operator op1, Operator op2) {
...
+  if exists(Class templ | c.isConstructedFrom(templ))
+  then // If we are an instantiation, then we can't check `op` if the
+       // template has it but we don't (because that member wasn't
+       // instantiated)
+       exists(Class templ | c.isConstructedFrom(templ) and
+                            (templ.getAMember().hasName(op) implies
+                             exists(Function f | f = c.getAMember() and
+                                                 f.hasName(op) and
+                                                 f.isDefined())))
+  else any()
```
Comment (geoffw0) [inline cpp/ql/src/jsf/4.10 Classes/AV Rule 85.ql]: Would this be more readable without the `if`, as `exists(Class temp1 | c.isConstructedFrom(temp1) and ...) or not exists(Class temp1 | c.isConstructedFrom(temp1))`?

**Comment by ian-semmle** [comment]
174f1fe was PR 23883, which has description "It looks like the query is doing the wrong thing for templates, but this just adds a test of the status quo for now.".

I've changed it to look at the template bodies instead of the instantiation bodies. Note that this means allowing a `LTExpr` rather than a `FunctionCall` to `operator<`.

File: cpp/ql/src/jsf/4.10 Classes/AV Rule 85.ql (hunk 2 above)
Comment (ian-semmle) [reply cpp/ql/src/jsf/4.10 Classes/AV Rule 85.ql]: I changed it to be an `implies`

File: cpp/ql/test/query-tests/jsf/4.10 Classes/AV Rule 85/AV Rule 85.expected (hunk 1 above)
Comment (ian-semmle) [reply cpp/ql/test/query-tests/jsf/4.10 Classes/AV Rule 85/AV Rule 85.expected]: These cases really are just testing the various cases where some things are and aren't used. Although the current query definition means that several of them get no alerts, I think they're still useful test cases for potential future changes to the query.

**Pull request merged**